### PR TITLE
feat(components): add Jungle Grid workload execution bundle

### PR DESCRIPTION
**Pull request by dejaguarkyng**
## Summary

Adds a production-ready Jungle Grid bundle for asynchronous AI workload execution from Langflow. The components use the Jungle Grid REST API directly through a shared client and retain Langflow's native bundle, component, icon, lazy-loading, documentation, and test conventions.

## Components

- Estimate Job
- Submit Job
- Create Job Input Upload
- List Job Inputs
- List Jobs
- Get Job Status
- Get Job Events
- Get Job Runtime
- Get Job Logs
- Cancel Job
- List Job Artifacts
- Create Artifact Download URL

## Production workflow

Estimate -> prepare managed inputs or scripts -> submit -> inspect status and lifecycle events -> poll paginated logs -> retrieve artifacts.

Upload-slot creation returns the temporary upload and completion information needed to transfer bytes outside the component. It does not claim that a file has been uploaded or completed.

## Current API alignment

- Supports command arrays and the legacy command-plus-args representation.
- Supports canonical `inference`, `training`, `fine_tuning`, and `batch` workload types, with REST normalization for fine-tuning.
- Supports environment objects, managed `input_files` and `script_files`, expected artifacts, routing, templates, metadata, and verified callback fields.
- Adds managed input upload slots, uploaded-input listing, cursor-paginated job listing, and platform lifecycle events.
- Preserves execution phases, delayed-start information, scheduling, routing, failure, cost, and artifact-contract fields returned by status responses.
- Keeps lifecycle events, runtime diagnostics, and cursor-paginated workload logs as distinct operations.
- Uses current job, runtime, cancellation, artifact-listing, and artifact-download endpoints.

## Safety

- Estimates do not submit jobs, reserve capacity, start compute, or guarantee startup.
- Submission may consume Jungle Grid credits; cancellation remains an explicit component.
- API credentials remain server-side and are sent only in the Bearer authorization header.
- Base URLs must be path-free HTTPS origins with a hostname.
- The shared client recursively redacts credentials, environment data, upload tokens, and temporary URLs from status and error text.
- Temporary upload and artifact URLs are returned only through intended outputs and are not stored in component status or logs.
- Side-effecting submission, cancellation, and signed-URL operations are not automatically retried.
- Tests mock every Jungle Grid request; CI does not start paid workloads.

## Backward compatibility

- Existing component class identifiers remain available.
- Legacy command strings plus `args` remain supported without changing execution semantics.
- Existing API credential inputs remain valid.
- `workload_name` replaces the colliding serialized input field while legacy serialized `name` values are read as a compatibility fallback.
- The legacy workload field and single `script_file` input-ID alias remain accepted.

## Review fixes

- Ran Ruff and corrected test import ordering.
- Replaced broad exception assertions with specific `JungleGridError` matches.
- Renamed behavior tests to describe what they verify.
- Ensured external icon props cannot override the controlled theme stroke.
- Added parent and bundle public exports plus discovery coverage.
- Hardened API base URL validation and normalization.
- Preserved explicitly supplied empty arrays and objects.
- Removed the workload-name/component-name collision.
- Redacted signed upload and artifact URLs from component status and free-form API errors.
- Added useful public docstrings and expanded component-specific coverage.

## Testing

Passed locally:

```text
/tmp/jungle-grid-test-venv/bin/ruff check src/lfx/src/lfx/components/jungle_grid src/backend/tests/unit/components/bundles/jungle_grid
All checks passed

/tmp/jungle-grid-test-venv/bin/ruff format --check src/lfx/src/lfx/components/jungle_grid src/backend/tests/unit/components/bundles/jungle_grid
17 files already formatted

/tmp/jungle-grid-test-venv/bin/ruff check --select D --ignore D100,D104,D107 src/lfx/src/lfx/components/jungle_grid
All checks passed

/tmp/jungle-grid-test-venv/bin/pytest --confcutdir=src/backend/tests/unit/components/bundles/jungle_grid src/backend/tests/unit/components/bundles/jungle_grid/test_jungle_grid.py -q
67 passed

/tmp/jungle-grid-test-venv/bin/python -m compileall -q src/lfx/src/lfx/components/jungle_grid src/backend/tests/unit/components/bundles/jungle_grid
Passed

Bundle discovery/import assertion
12 Jungle Grid components discovered; parent and bundle exports passed

/tmp/jg-biome/node_modules/.bin/biome check src/icons/JungleGrid/JungleGridIcon.jsx src/icons/JungleGrid/index.tsx src/icons/JungleGrid/__tests__/jungle-grid-icon.test.tsx src/icons/lazyIconImports.ts src/utils/styleUtils.ts
Checked 5 files; no fixes required

/tmp/jg-jest/node_modules/.bin/jest src/icons/JungleGrid/__tests__/jungle-grid-icon.test.tsx --runInBand --config /tmp/jg-jest.config.cjs
1 suite passed; 2 tests passed

git diff --check origin/release-1.10.0...HEAD
Passed

Focused credential and signed-URL scan of changed files
Passed; no embedded credentials or temporary URLs
```

The full workspace dependency sync could not complete locally because extracting optional Torch/OpenCV packages exceeded the runner's `/tmp` quota. Full frontend/docs dependency installation also encountered registry timeouts, so the focused frontend checks above use isolated repository-pinned tools.

Fresh GitHub CI results:

- Ruff style and autofix checks: passed
- Component index update: passed
- Backend unit matrix on Python 3.10 and 3.14: passed
- Backend integration matrix: passed
- LFX and CLI tests: passed
- Frontend Biome, Jest, and all Playwright shards: passed
- Documentation build: passed
- Starter templates and Docker image tests: passed
- Coverage merge and all other applicable code checks: passed
- Smoke/docs-draft/localization jobs: skipped by workflow conditions
- Aggregate `CI Success`: blocked only because the repository-wide nightly PyPI package was not updated on June 12, 2026 (`check-nightly-status.should-proceed=false`). Its log reports no failed or cancelled PR jobs.

## Known limitations

- Create Job Input Upload creates the managed slot only. Langflow does not expose arbitrary host paths or place binary file bytes in text inputs, so the caller must upload and complete the file using the temporary response data before submitting its `input_id`.

### CODE REVIEW

File: src/backend/tests/unit/components/bundles/jungle_grid/test_jungle_grid.py [hunk 1]
```diff
@@ -0,0 +1,285 @@
...
+
+from lfx.components.jungle_grid.artifact_download_url import JungleGridCreateArtifactDownloadURLComponent
+from lfx.components.jungle_grid.cancel_job import JungleGridCancelJobComponent
+from lfx.components.jungle_grid.estimate_job import JungleGridEstimateJobComponent
+from lfx.components.jungle_grid.get_job_logs import JungleGridGetJobLogsComponent
+from lfx.components.jungle_grid.get_job_runtime import JungleGridGetJobRuntimeComponent
+from lfx.components.jungle_grid.get_job_status import JungleGridGetJobStatusComponent
+from lfx.components.jungle_grid.list_job_artifacts import JungleGridListJobArtifactsComponent
+from lfx.components.jungle_grid.submit_job import JungleGridSubmitJobComponent
+from tests.base import ComponentTestBaseWithoutClient
```
[the remark this answers is not in the document]
Comment (dejaguarkyng) [reply src/backend/tests/unit/components/bundles/jungle_grid/test_jungle_grid.py]: Resolved. Imports are Ruff-sorted, and the focused Ruff check now passes.

File: src/backend/tests/unit/components/bundles/jungle_grid/test_jungle_grid.py [hunk 2]
```diff
@@ -0,0 +1,285 @@
...
+    async def test_http_errors_are_sanitized(self, component_class, default_kwargs, status_code: int) -> None:
+        component = component_class(**self._kwargs_for(component_class, default_kwargs))
+        method = getattr(component, component.outputs[0].method)
+        mocked = AsyncMock(return_value=_response(status_code, {"error": SECRET, "url": SIGNED_URL}))
+        with patch("httpx.AsyncClient.request", mocked), pytest.raises(ValueError) as exc_info:
+            await method()
+        error_text = str(exc_info.value)
+        assert f"Jungle Grid API error {status_code}" in error_text
+        assert SECRET not in error_text
+        assert SIGNED_URL not in error_text
```
[the remark this answers is not in the document]
Comment (dejaguarkyng) [reply src/backend/tests/unit/components/bundles/jungle_grid/test_jungle_grid.py]: Resolved. Exception assertions now use JungleGridError with behavior-specific match expressions, including API, timeout, network, and JSON validation cases.

File: src/frontend/src/icons/JungleGrid/JungleGridIcon.jsx [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+  <svg
+    xmlns="http://www.w3.org/2000/svg"
+    viewBox="0 0 24 24"
+    fill="none"
+    stroke={props.isdark === "true" ? "#9BFEAA" : "#1A0250"}
+    strokeWidth={2}
+    strokeLinecap="round"
+    strokeLinejoin="round"
+    {...props}
+  >
```
[the remark this answers is not in the document]
Comment (dejaguarkyng) [reply src/frontend/src/icons/JungleGrid/JungleGridIcon.jsx:11]: Resolved. General SVG props are spread before the controlled theme stroke, and the wrapper also applies ref/theme after general props. A frontend regression test covers both orderings.

File: src/lfx/src/lfx/components/__init__.py [hunk 4]
```diff
@@ -156,6 +157,7 @@
     "huggingface": "__module__",
     "icosacomputing": "__module__",
     "input_output": "__module__",
+    "jungle_grid": "__module__",
```
[the remark this answers is not in the document]
Comment (dejaguarkyng) [reply src/lfx/src/lfx/components/__init__.py]: Resolved. jungle_grid is present in the parent import map and __all__; discovery tests also verify __dir__ and all 12 bundle exports.

File: src/lfx/src/lfx/components/jungle_grid/_client.py [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+    """Sanitized Jungle Grid component error."""
+
+
+def normalize_base_url(api_base_url: str | None) -> str:
+    base_url = (api_base_url or DEFAULT_API_BASE_URL).strip().rstrip("/")
+    parsed = urlparse(base_url)
+    if parsed.scheme != "https" or not parsed.netloc:
+        msg = "Jungle Grid API base URL must be a valid HTTPS URL."
+        raise JungleGridError(msg)
+    return base_url
```
[the remark this answers is not in the document]
Comment (dejaguarkyng) [reply src/lfx/src/lfx/components/jungle_grid/_client.py:64]: Resolved. Base URL normalization now requires a hostname and path-free HTTPS origin and rejects credentials, paths, params, queries, and fragments. The cases are covered by focused tests.

File: src/lfx/src/lfx/components/jungle_grid/_client.py [hunk 6]
```diff
@@ -0,0 +1,180 @@
...
+    payload: dict[str, Any] = {
+        "name": require_text(name, "Name"),
+        "image": require_text(image, "Image"),
+        "workload_type": require_text(workload_type, "Workload Type"),
+        "model_size_gb": model_size,
+    }
+    if command_value := optional_text(command):
+        payload["command"] = command_value
+    if args_value := parse_json_field(args, "Args", list):
+        payload["args"] = args_value
```
[the remark this answers is not in the document]
Comment (dejaguarkyng) [reply src/lfx/src/lfx/components/jungle_grid/_client.py]: Resolved. Optional arrays and objects use explicit None checks, preserving args: [], metadata: {}, callback_metadata: {}, and managed-input empty arrays.

File: src/lfx/src/lfx/components/jungle_grid/_shared.py [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+
+def workload_inputs() -> list:
+    return [
+        MessageTextInput(
+            name="name",
+            display_name="Name",
+            required=True,
+            tool_mode=True,
+            info="Name for the Jungle Grid workload.",
+        ),
```
[the remark this answers is not in the document]
Comment (dejaguarkyng) [reply src/lfx/src/lfx/components/jungle_grid/_shared.py]: Resolved. Submit Job now exposes workload_name and maps it to API name, with a compatibility fallback for existing serialized name values.

[the remark this answers is not in the document]
Comment (dejaguarkyng) [reply src/lfx/src/lfx/components/jungle_grid/_shared.py]: Resolved. Submit Job now exposes workload_name and maps it to API name, with a compatibility fallback for existing serialized name values.

File: src/lfx/src/lfx/components/jungle_grid/artifact_download_url.py [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+    ]
+    outputs = [Output(display_name="JSON", name="data", method="create_artifact_download_url")]
+
+    async def create_artifact_download_url(self) -> Data:
+        job_id = path_segment(self.job_id, "Job ID")
+        artifact_id = path_segment(self.artifact_id, "Artifact ID")
+        client = JungleGridClient(self.api_key, self.api_base_url)
+        result = await client.request("POST", f"/v1/jobs/{job_id}/artifacts/{artifact_id}/download", json_body={})
+        data = Data(data=result)
+        self.status = data
```
[the remark this answers is not in the document]
Comment (dejaguarkyng) [reply src/lfx/src/lfx/components/jungle_grid/artifact_download_url.py]: Resolved. The real temporary artifact download response is returned only from the output; component status stores a redacted summary. Upload-slot status follows the same rule.

File: src/lfx/src/lfx/components/jungle_grid/submit_job.py [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+            command=self.command,
+            args=self.args,
+            optimize_for=self.optimize_for,
+        )
+        if callback_url := optional_text(self.callback_url):
+            payload["callback_url"] = callback_url
+        if callback_auth_token := optional_text(self.callback_auth_token):
+            payload["callback_auth_token"] = callback_auth_token
+        if callback_metadata := parse_json_field(self.callback_metadata, "Callback Metadata", dict):
+            payload["callback_metadata"] = callback_metadata
```
[the remark this answers is not in the document]
Comment (dejaguarkyng) [reply src/lfx/src/lfx/components/jungle_grid/submit_job.py]: Resolved. Explicit empty callback metadata is preserved with an is-not-None check and covered by the submission payload test.

**Comment by dejaguarkyng** [comment]
Thanks for the review and detailed feedback.

I updated the existing Jungle Grid branch against the latest `release-1.10.0` base and resolved the outstanding findings around import ordering, specific exception assertions, URL validation, bundle exports, component field collisions, explicit empty values, icon theming, signed URL handling, docstrings, and test coverage.

The bundle is now aligned with the current Jungle Grid production workflow, including list jobs, managed input upload slots, uploaded-input listing, lifecycle events, command-array submissions, input/script references, expected artifacts, and updated status, runtime, log, cancellation, and artifact handling.

I re-ran the focused Ruff, pytest, bundle discovery, frontend Biome/Jest, bytecode, diff, and secret checks. The exact local results and full-workspace CI note are included in the PR description. The PR is ready for another review.

**Comment by dejaguarkyng** [comment]
@coderabbitai review

**Comment by dejaguarkyng** [comment]
@coderabbitai The reported duplicate `Develop/observability-grafana-loki` sidebar ID is pre-existing and unchanged in `origin/release-1.10.0` at the same two locations. This PR changes `docs/sidebars.js` only to add `Components/bundles-jungle-grid` under the bundle list, so I am leaving the unrelated base-branch duplication untouched to keep the diff scoped.
